test/e2e/io_soak/io_soak_test.go

```diff
@@ -0,0 +1,212 @@
+	argThinkTime := fmt.Sprintf("--thinktime=%d", tt)
+	argThinkTimeBlocks := fmt.Sprintf("--thinktime_blocks=%d", ttb)
+	cmd := exec.Command(
+		"kubectl",
+		"exec",
+		"-it",
+		podName,
+		"--",
+		"fio",
+		"--name=benchtest",
+		"--size=50m",
+		"--filename=/volume/test",
+		"--direct=1",
+		"--rw=randrw",
+		"--ioengine=libaio",
+		"--bs=4k",
+		"--iodepth=16",
+		"--numjobs=1",
+		"--time_based",
+		argThinkTime,
+		argThinkTimeBlocks,
+		argRuntime,
+	)
+	cmd.Dir = ""
+	output, err := cmd.CombinedOutput()
+	logf.Log.Info("Finished running fio", "pod", podName, "duration", duration)
+	logf.Log.Info(string(output))
+	if err != nil {
+		errC <- err
+	} else {
+		doneC <- podName
+	}
+}
+
+func makeVols(scName string, count int) {
+	for ix := 1; ix <= count; ix++ {
+		volName := fmt.Sprintf("%s-%d", scName, ix)
+		common.MkPVC(volName, scName)
+		tmp := volSc{volName, scName}
+		volNames = append(volNames, tmp)
+	}
+}
+
+func rmVols(scName string, count int) {
+	for ix := 1; ix <= count; ix++ {
+		volName := fmt.Sprintf("%s-%d", scName, ix)
+		common.RmPVC(volName, scName)
+	}
+}
+
+/// proto - protocol "nvmf" or "isci"
+/// replicas - number of replicas for each volume
+/// loadFactor - number of volumes for each mayastor instance
+func IOSoakTest(proto string, replicas int, loadFactor int, duration int) {
+
+	nodeList, err := common.GetNodeLocs()
+	Expect(err).ToNot(HaveOccurred())
+
+	num_mayastor_nodes := 0
+	jobCount := 0
+	sort.Slice(nodeList, func(i, j int) bool { return nodeList[i].NodeName < nodeList[j].NodeName })
+	for i, node := range nodeList {
+		if node.MayastorNode && !node.MasterNode {
+			logf.Log.Info("MayastorNode", "name", node.NodeName, "index", i)
+			jobCount += loadFactor
+			num_mayastor_nodes += 1
+		}
+	}
+
+	Expect(replicas <= num_mayastor_nodes).To(BeTrue())
+
+	scName := fmt.Sprintf("io-soak-%s", proto)
+	common.MkStorageClass(scName, replicas, proto, "io.openebs.csi-mayastor")
+	scNames = append(scNames, scName)
+
+	logf.Log.Info("IOSoakTest", "jobs", jobCount, "volumes", jobCount, "test pods", jobCount)
+
+	// Create the required set of volumes
+	makeVols(scName, jobCount)
+    doneC, errC := make(chan string), make(chan error)
+
+    // Create the test pods for the volumes
+	for _, vol := range volNames {
+		// Create the fio Pod
+		fioPodName := "fio-" + vol.volName
+		pod, err := common.CreateFioPod(fioPodName, vol.volName)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(pod).ToNot(BeNil())
+		podNames = append(podNames, fioPodName)
+	}
+
+	// Wait for the test pods to be ready
+	for _, podName := range podNames  {
+		// Wait for the fio Pod to transition to running
+		Eventually(func() bool {
+			return common.IsPodRunning(podName)
+		},
+			defTimeoutSecs,
+			"1s",
+		).Should(Equal(true))
+	}
+
+	// Execute the application in the pods concurrently.
+	for x, podName := range podNames  {
+		// Run the fio test
+		go runFio(podName, duration, x*500, x*100, doneC, errC)
```

Comment:
Maybe a comment to describe the thinktime parameters could be needed here.

Reply:
Done